### ISSUE TITLE
Issue utilising Free.fold 

### ISSUE THREAD

**Issue by Choc13** [fsprojects/FSharpPlus#405]
### Description

I'm trying to use `Free.fold` to write an interpreter for my free monad, but I'm struggling to satisfy the type checker with my implementation of the mapping function `f`. It's entirely possible that there is a problem with my code, but oddly I can satisfy the compiler (and run the code without errors) if I copy the definition of `Free.fold` into my project and remove all type hints from the signature. 

### Repro steps

I have a DSL and associated workflow like so:

```fsharp
open FSharpPlus
open FSharpPlus.Data

type FooId = FooId of string

type Foo = { Id: FooId }

module GetFoo =
    type Error = NotFound

    type Instruction<'next> =
        | Read of (FooId * (Foo option -> 'next))
        static member Map(instruction, f: 'a -> 'b) =
            match instruction with
            | Read (id, next) -> Read(id, next >> f)

    type Program<'a> = Free<Instruction<'a>, 'a>

    let read fooId = Read(fooId, id) |> Free.liftF

    type Request = { Id: FooId }
    type Response = Result<Foo, Error>

    let handle request: Program<Response> =
        monad {
            let! foo = read request.Id
            return foo |> Option.toResultWith NotFound
        }
```

Then, to keep things simple, in my tests I would like to write an interpreter which targets the identity monad. So I have written:

```fsharp
let interpreter: Program<Response> -> Response =
    Free.fold (function
        | Read (fooId, next) -> monad { return Some({Id = fooId}) |> next })
    >> Identity.run
```

### Expected behavior

I would expect this to compile

### Actual behavior

But it produces the following compiler error:

```
Type mismatch. Expecting a
    'Program<Response> -> 'a'    
but given a
    'Free<Instruction<Response>,Response> -> 'b'    
Type mismatch. Expecting a
    'Instruction<Response>'    
but given a
    'Instruction<Free<Instruction<Response>,Response>>'    
The type 'Response' does not match the type 'Free<Instruction<Response>,Response>'
```

Am I missing a type hint somewhere, or is my implementation of the mapping function from `Instruction` to `Identity` monad fundamentally wrong somehow? 

### Known workarounds

Currently I can get this to compile by redefining `fold` without the type hints in the signature:

```fsharp
open FSharpPlus
open FSharpPlus.Data

[<RequireQualifiedAccess>]
module Free =
    let inline fold2 f x =
        let rec loop f x =
            match Free.run x with
            | Pure a -> monad { return a }
            | Roll x -> f x >>= loop f

        loop f x
```

And then switching the above implementation of the interpreter to use `Free.fold2`.

### Related information 

* Operating system - Mac OSX
* FSharpPlus Version - 1.1.5
* .NET Runtime - .NET 5.0
* F# lang version - 5

**Comment by gusty** [comment]
Thanks for reporting. I'll try to understand what's happening here.
Could you edit your repro and add the missing types: `Foo` and `FooId` ?

**Comment by Choc13** [comment]
Hi @gusty thanks for picking this up. I'll edit the repro with those types now.

**Comment by gusty** [comment]
Thanks.
I tried your workaround with `fold2` but this is what I get:

```
    let interpreter: Program<Response> -> Response =
        Free.fold2 (function
            | Read (fooId, next) -> result ( Some({Id = fooId}) |> next) )
        >> Identity.run
;;

              | Read (fooId, next) -> result ( Some({Id = fooId}) |> next) )
  -------------------------------------------------------------------^^^^

stdin(7,68): error FS0001: Type mismatch. Expecting a
    'Request option -> 'a'    
but given a
    'Foo option -> Free<Instruction<Response>,Response>'    
The type 'Request' does not match the type 'Foo'
```

Are you sure it works for you?

**Comment by Choc13** [comment]
Ah sorry, it was because I redacted some of the code and in doing so `Request` and `Foo` became isomorphic without a typehint. So if you change it to:

```fsharp
let interpreter: Program<Response> -> Response =
        Free.fold2 (function
            | Read (fooId, next) -> monad { return Some({ Id = fooId }: Foo) |> next })
        >> Identity.run
```

Then it compiles for me.

**Comment by gusty** [comment]
This is a bug, the `fold` signature is wrong.

I think we can define it as `let inline fold (f: '``Functor<Free<'Functor<'T>,'T>>`` -> '``Monad<Free<'Functor<'T>,'T>>>``) (x: Free<'``Functor<'T>``,'T>) : '``Monad<'T>`` = ...`

Also, we don't have any test for `fold`, which would have prevented this.

Do you think you can submit a PR with this fix and include your code as a test?

**Comment by Choc13** [comment]
Ah yes of course that makes sense! The current type signature seems to make sense, but the mapping is still to another `Free<Functor>` at that stage, as it's only when we call the recursive run and finally reach the pure case does it map it to the output value of the process. I'll open a PR with some test cases when I get a bit of time later! Thanks for solving this.

**Comment by gusty** [comment]
Actually the current signature is not that bad, the error is that is uses always `'T` whereas `f` should use a different type parameter, otherwise, it unifies its result `Monad<'T>` with the result of `f`.

If we fix that, we end up with:

`let inline fold (f: '``Functor<'T>`` -> '``Monad<'T>``) (x: Free<'``Functor<'U>``,'U>) : '``Monad<'U>`` = ...`

which is more generic than the one I proposed, because `Free<..>` is a `Functor<..>` and a `Monad<..>`.

Now, tbh I still didn't came across a case where you supply an `f` with a `Functor` other than `Free`. It would be interesting to get one for our tests.

**Comment by Choc13** [comment]
Ah yeah great point that’s exactly it! I can’t think of a situation either
but having a more flexible signature seems like the best way to go here, so
I’ll implement it like you’ve suggested.

Le mar. 1 déc. 2020 à 15:58, Gustavo Leon <notifications@github.com> a
écrit :

**Comment by Choc13** [comment]
@gusty thanks for the tip regarding `result x` instead of `monad { return x }`.

**Comment by gusty** [comment]
Fixed by #406
